apps/recnet-api/src/modules/rec/rec.service.ts

```diff
@@ -0,0 +1,259 @@
+    }
+    // article and articleId cannot be null at the same time
+    // check libs/recnet-api-model/src/lib/api/rec.ts
+    if (!article && !articleId) {
+      throw new RecnetError(
+        ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+        HttpStatus.BAD_REQUEST,
+        "Article and articleId cannot be null at the same time"
+      );
+    }
+
+    let targetArticleId: string | null = null;
+    if (article) {
+      if (articleId) {
+        throw new RecnetError(
+          ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+          HttpStatus.BAD_REQUEST,
+          "Article and articleId cannot have value at the same time"
+        );
+      }
+      // check if article already exists
+      const dbArticle = await this.articleRepository.findArticleByLink(
+        article.link
+      );
+      if (dbArticle) {
+        targetArticleId = dbArticle.id;
+      } else {
+        const newArticle = await this.articleRepository.createArticle(article);
+        targetArticleId = newArticle.id;
+      }
+    }
+
+    const articleIdToConnect = articleId ? articleId : targetArticleId;
+    if (articleIdToConnect === null) {
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.INTERNAL_SERVER_ERROR,
+        "Error getting article id while creating rec"
+      );
+    }
+    const newRec = await this.recRepository.createRec(
+      userId,
+      description,
+      articleIdToConnect
+    );
+    return {
+      rec: this.getRecFromDbRec(newRec),
+    };
+  }
+
+  public async updateUpcomingRec(
+    articleId: string | null,
+    article: CreateArticleInput | null,
+    description: string,
+    userId: string
+  ): Promise<UpdateRecResponse> {
+    const dbRec = await this.recRepository.findUpcomingRec(userId);
+    if (!dbRec) {
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.NOT_FOUND,
+        "Upcoming rec not found"
+      );
+    }
+    // article and articleId cannot be null at the same time
+    // check libs/recnet-api-model/src/lib/api/rec.ts
+    if (!article && !articleId) {
+      throw new RecnetError(
+        ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+        HttpStatus.BAD_REQUEST,
+        "Article and articleId cannot be null at the same time"
+      );
+    }
+
+    let targetArticleId: string | null = null;
+    if (article) {
+      if (articleId) {
+        throw new RecnetError(
+          ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+          HttpStatus.BAD_REQUEST,
+          "Article and articleId cannot have value at the same time"
+        );
+      }
+      // check if article already exists
+      const dbArticle = await this.articleRepository.findArticleByLink(
+        article.link
+      );
+      if (dbArticle) {
+        targetArticleId = dbArticle.id;
+      } else {
+        const newArticle = await this.articleRepository.createArticle(article);
+        targetArticleId = newArticle.id;
+      }
+    }
```

Comment:
[nit] This part could be extracted to a private method such as `findOrCreateTargetArticleId`.

Reply:
I've thought about it when I wrote it but I think doing two different things in a function looks weird to me (violate single responsibility?). Sometimes I think DRY(Don't repeat yourself) is a little over-abstracted.

I can change it later if you want it tho